### PR TITLE
Prelude to #7388 - dependency bump for nlohmann json and valijson

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

Prelude to #7388 - dependency bump for nlohmann json and valijson

* valijson bumped to current master (https://github.com/tristanpenman/valijson/commit/92d369e91605c73c32587fb777738f17d3e23405), I could have bumped only to v0.2 I suppose, but reviewing the 5 commits to master snce v0.2 I see no reason no to use recent ones
* While I was at it (and early on) I also bumped nlohmann's json from 3.7.0 to 3.7.3
* Create a patch file for both, and filled the third_party/README.md with info
* Update the InputValidation to match the new output and updated on test that checks for a RunPeriod with no name since it is affected.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: third_party/valijson/valijson/compat/optional.hpp [hunk 1]
```diff
@@ -0,0 +1,1042 @@
+// Copyright (C) 2011 - 2012 Andrzej Krzemienski.
```
Comment (jmarrec) [inline third_party/valijson/valijson/compat/optional.hpp:1]: Not sure why this file was non-existing in the current E+ develop and whether it should have been comitted here.

File: third_party/valijson/valijson/validation_visitor.hpp [hunk 2]
```diff
@@ -874,9 +874,16 @@ class ValidationVisitor: public constraints::ConstraintVisitor
...
+                    for (const typename AdapterType::ObjectMember m : object) {
+                        if (propertiesMatched.find(m.first) == propertiesMatched.end()) {
+                            unwanted = m.first;
+                            break;
+                        }
+                    }
+                    results->pushError(context, "Object contains a property "
                             "that could not be validated using 'properties' "
-                            "or 'additionalProperties' constraints");
+                            "or 'additionalProperties' constraints: '" + unwanted + "'.");
```
Comment (jmarrec) [inline third_party/valijson/valijson/validation_visitor.hpp:886]: This is the main diff in valijson's recent commits that I am looking for.

File: tst/EnergyPlus/unit/RunPeriod.unit.cc [hunk 3]
```diff
@@ -299,7 +299,7 @@ TEST_F(EnergyPlusFixture, RunPeriod_NoName) {
 
     std::string const error_string =
         delimited_string({
-            "   ** Severe  ** <root>[RunPeriod] - Object contains properties that could not be validated using 'properties' or 'additionalProperties' constraints",
+            "   ** Severe  ** <root>[RunPeriod] - Object contains a property that could not be validated using 'properties' or 'additionalProperties' constraints: ''.",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/RunPeriod.unit.cc:302]: Update Unit test to match

File: src/EnergyPlus/InputProcessing/InputValidation.cc [hunk 4]
```diff
@@ -102,7 +102,7 @@ bool Validation::validate(json const &parsed_input)
                 errors_.emplace_back(context + " - " + error.description);
                 if (max_context == 2 &&
                     error.description ==
-                        "Object contains properties that could not be validated using 'properties' or 'additionalProperties' constraints") {
+                        "Object contains a property that could not be validated using 'properties' or 'additionalProperties' constraints: ''.") {
```
Comment (jmarrec) [inline src/EnergyPlus/InputProcessing/InputValidation.cc:105]: Update InputValidation condition to match valijson's new output message.

File: third_party/nlohmann/json373.patch [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
++    using object_comparator_t = doj::alphanum_less<>;
+ #else
+-    using object_comparator_t = std::less<StringType>;
++    using object_comparator_t = doj::alphanum_less<StringType>;
+ #endif
+ 
+     /*!
+-- 
+2.26.2
+
```
Comment (jmarrec) [inline third_party/nlohmann/json373.patch:36]: Actual patch file for nlohmann's json.

File: third_party/valijson/valijson/valijson.patch [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+From: Julien Marrec <julien.marrec@gmail.com>
+Date: Tue, 19 May 2020 15:54:51 +0200
+Subject: [PATCH] Repatch like current develop
+
+---
+ .../valijson/valijson/validation_visitor.hpp  | 57 ++++++++++++++-----
+ 1 file changed, 44 insertions(+), 13 deletions(-)
+ mode change 100755 => 100644 third_party/valijson/valijson/validation_visitor.hpp
+
+diff --git a/third_party/valijson/valijson/validation_visitor.hpp b/third_party/valijson/valijson/validation_visitor.hpp
```
Comment (jmarrec) [inline third_party/valijson/valijson/valijson.patch:11]: Actual patch file to valijson.

**Comment by Myoldmopar** [comment]
I've personally got no problem with these changes.  CI has no problem with these changes (spurious known Mac test failure and Windows MTD failure that may be fixed with #7997).  I'd love for @mbadams5 to give this an approved status, if he approves of course.

**Comment by mbadams5** [comment]
My only comment is to probably pull the latest develop of the JSON library instead of 3.7.3, especially since we'll probably update again before the next EnergyPlus release.

**Comment by Myoldmopar** [comment]
I'm not sure what develop offers over 3.7.3 so I can't comment there.  If it is worth doing develop here instead of 3.7.3, then make it so and we'll get this merged.

**Comment by jmarrec** [comment]
@Myoldmopar @mbadams5  I pulled latest nlohmann json and merged latest E+ develop in (for reference: from https://github.com/nlohmann/json/blob/88292d949939b4f1b0dbf8a11fb39db6a7e91e15/single_include/nlohmann/json.hpp)

File: third_party/nlohmann/json373.patch (hunk 5 above)
Comment (Myoldmopar) [reply third_party/nlohmann/json373.patch:36]: :+1:

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
Works for me!  Merging.  Thanks @jmarrec.  I'll look at #8006 shortly.

**Pull request merged**